bindings/declarations.go

```diff
@@ -69,23 +64,41 @@ func (p *TypeParameter) isNode() {}
 
 // Simplify removes duplicate type parameters
 func Simplify(p []*TypeParameter) ([]*TypeParameter, error) {
-	params := make([]*TypeParameter, 0, len(p))
-	exists := make(map[string]*TypeParameter)
+	params := []*TypeParameter{}
+	set := make(map[string]bool)
 	for _, tp := range p {
-		if found, ok := exists[tp.Name.Ref()]; ok {
-			// Compare types, make sure they are the same
-			equal := reflect.DeepEqual(found, tp)
-			if !equal {
-				return nil, fmt.Errorf("type parameter %q already exists with different type", tp.Name)
+		ref := tp.Name.Ref()
+		if _, ok := set[ref]; !ok {
+			params = append(params, tp)
+			set[ref] = true
+
+			if union, ok := tp.Type.(*UnionType); ok {
+				simplifyUnion(union)
 			}
-			continue
 		}
-		params = append(params, tp)
-		exists[tp.Name.Ref()] = tp
 	}
 	return params, nil
 }
 
+func simplifyUnion(union *UnionType) *UnionType {
+	types := []ExpressionType{}
+	set := map[string]bool{}
```

Comment:
Since the set is only tracking literals, can we rename this? If we eventually add deduping more complex types, I think we will need to come up with some other compare mechanism


```suggestion
	literalsSet := map[string]bool{}
```